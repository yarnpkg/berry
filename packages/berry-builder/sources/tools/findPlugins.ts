import {UsageError} from 'clipanion';

export function findPlugins({basedir, profile, plugins}: {basedir: string, profile: string, plugins: Array<string>}) {
  const pkgJson = require(`${basedir}/package.json`);

  if (!pkgJson[`@berry/builder`] || !pkgJson[`@berry/builder`].bundles)
    throw new UsageError(`This command requires your package.json to contain specific configuration keys`);

  if (!profile)
    profile = `standard`;

  const profiles = profile.split(/\+/g);
  const supported = new Set(Object.keys(pkgJson[`@berry/builder`].bundles));

  if (profiles.some(profile => !supported.has(profile)))
    throw new UsageError(`Invalid profile "${profile}", expected one of ${[...supported].join(`, `)}`);

  return Array.from(new Set(profiles.reduce((acc, profile) => {
    return acc.concat(pkgJson[`@berry/builder`].bundles[profile]);
  }, plugins)));
}
