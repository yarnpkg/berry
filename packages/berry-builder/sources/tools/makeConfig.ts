// @ts-ignore
import PnpWebpackPlugin from 'pnp-webpack-plugin';
import merge            from 'webpack-merge';
import webpack          from 'webpack';

export const makeConfig = (config: webpack.Configuration) => merge({
  mode: `none`,
  devtool: false,

  target: `node`,

  node: {
    __dirname: false,
    __filename: false,
  },

  output: {
    libraryTarget: `commonjs2`,
  },

  resolve: {
    alias: {
      [`supports-color`]: `supports-color/index`,
      [`agentkeepalive`]: `agentkeepalive/index`,
    },
    extensions: [`.js`, `.ts`, `.tsx`, `.json`],
    mainFields: [`browser`, `module`, `main`],
    plugins: [PnpWebpackPlugin],
  },

  module: {
    rules: [{
      test: /\.tsx?$/,
      exclude: /\.d\.ts$/,
      loader: require.resolve(`ts-loader`),
      options: PnpWebpackPlugin.tsLoaderOptions({
        compilerOptions: {
          declaration: false,
        },
      }),
    }],
  },

  externals: {
    // Both of those are native dependencies of text-buffer we can't bundle
    [`fs-admin`]: `{}`,
    [`pathwatcher`]: `{}`,
  },

  plugins: [
    new webpack.IgnorePlugin(/^encoding$/, /node-fetch/),
    new webpack.DefinePlugin({[`IS_WEBPACK`]: `true`}),
  ],
}, config);
