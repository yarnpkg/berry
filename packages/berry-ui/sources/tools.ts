export function iterate(source: any, cb: (value: any, index: number, virtualIndex: number) => any, ifEmpty?: any) {
  const elements = [];
  let index = 0;

  for (const item of source) {
    const element = cb(item, index++, elements.length);

    if (element) {
      elements.push(element);
    }
  }

  if (elements.length === 0)
    return ifEmpty;

  return elements;
}
