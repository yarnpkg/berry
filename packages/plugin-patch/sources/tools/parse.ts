import {miscUtils}           from '@yarnpkg/core';
import {PortablePath, npath} from '@yarnpkg/fslib';

const HEADER_REGEXP = /^@@ -(\d+)(,(\d+))? \+(\d+)(,(\d+))? @@.*/;

export type HunkHeader = {
  original: {
    start: number,
    length: number,
  },
  patched: {
    start: number,
    length: number,
  },
};

export function parseHunkHeaderLine(headerLine: string): HunkHeader {
  const match = headerLine.trim().match(HEADER_REGEXP);
  if (!match)
    throw new Error(`Bad header line: '${headerLine}'`);

  return {
    original: {
      start: Math.max(Number(match[1]), 1),
      length: Number(match[3] || 1),
    },
    patched: {
      start: Math.max(Number(match[4]), 1),
      length: Number(match[6] || 1),
    },
  };
}

export const NON_EXECUTABLE_FILE_MODE = 0o644;
export const EXECUTABLE_FILE_MODE = 0o755;

type FileMode =
  | typeof NON_EXECUTABLE_FILE_MODE
  | typeof EXECUTABLE_FILE_MODE;

type PatchMutationPart = {
  type: `context` | `insertion` | `deletion`,
  lines: Array<string>,
  noNewlineAtEndOfFile: boolean,
};

type FileRename = {
  type: `rename`,
  semverExclusivity: string | null,
  fromPath: PortablePath,
  toPath: PortablePath,
};

type FileModeChange = {
  type: `mode change`,
  semverExclusivity: string | null,
  path: PortablePath,
  oldMode: FileMode,
  newMode: FileMode,
};

export type FilePatch = {
  type: `patch`,
  semverExclusivity: string | null,
  path: PortablePath,
  hunks: Array<Hunk>,
  beforeHash: string | null,
  afterHash: string | null,
};

type FileDeletion = {
  type: `file deletion`
  semverExclusivity: string | null,
  path: PortablePath,
  mode: FileMode,
  hunk: Hunk | null,
  hash: string | null,
};

type FileCreation = {
  type: `file creation`,
  semverExclusivity: string | null,
  mode: FileMode,
  path: PortablePath,
  hunk: Hunk | null,
  hash: string | null,
};

export type PatchFilePart =
  | FilePatch
  | FileDeletion
  | FileCreation
  | FileRename
  | FileModeChange;

export type ParsedPatchFile =
  Array<PatchFilePart>;

type State =
  | `parsing header`
  | `parsing hunks`;

type FileDeets = {
  semverExclusivity: string | null,
  diffLineFromPath: string | null,
  diffLineToPath: string | null,
  oldMode: string | null,
  newMode: string | null,
  deletedFileMode: string | null,
  newFileMode: string | null,
  renameFrom: string | null,
  renameTo: string | null,
  beforeHash: string | null,
  afterHash: string | null,
  fromPath: string | null,
  toPath: string | null,
  hunks: Array<Hunk> | null,
};

export type Hunk = {
  header: HunkHeader,
  parts: PatchMutationPart[],
};

const emptyFilePatch = (): FileDeets => ({
  semverExclusivity: null,
  diffLineFromPath: null,
  diffLineToPath: null,
  oldMode: null,
  newMode: null,
  deletedFileMode: null,
  newFileMode: null,
  renameFrom: null,
  renameTo: null,
  beforeHash: null,
  afterHash: null,
  fromPath: null,
  toPath: null,
  hunks: null,
});

const emptyHunk = (headerLine: string): Hunk => ({
  header: parseHunkHeaderLine(headerLine),
  parts: [],
});

const hunkLinetypes: {[k: string]: PatchMutationPart['type'] | `pragma` | `header`} = {
  [`@`]: `header`,
  [`-`]: `deletion`,
  [`+`]: `insertion`,
  [` `]: `context`,
  [`\\`]: `pragma`,
  // Treat blank lines as context
  undefined: `context`,
};

function parsePatchLines(lines: Array<string>, {supportLegacyDiffs}: {supportLegacyDiffs: boolean}) {
  const result: Array<FileDeets> = [];

  let currentFilePatch: FileDeets = emptyFilePatch();
  let state: State = `parsing header`;
  let currentHunk: Hunk | null = null;
  let currentHunkMutationPart: PatchMutationPart | null = null;

  function commitHunk() {
    if (currentHunk) {
      if (currentHunkMutationPart) {
        currentHunk.parts.push(currentHunkMutationPart);
        currentHunkMutationPart = null;
      }

      currentFilePatch.hunks!.push(currentHunk);
      currentHunk = null;
    }
  }

  function commitFilePatch() {
    commitHunk();

    result.push(currentFilePatch);
    currentFilePatch = emptyFilePatch();
  }

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (state === `parsing header`) {
      if (line.startsWith(`@@`)) {
        state = `parsing hunks`;
        currentFilePatch.hunks = [];
        i -= 1;
      } else if (line.startsWith(`diff --git `)) {
        if (currentFilePatch && currentFilePatch.diffLineFromPath)
          commitFilePatch();

        const match = line.match(/^diff --git a\/(.*?) b\/(.*?)\s*$/);
        if (!match)
          throw new Error(`Bad diff line: ${line}`);

        currentFilePatch.diffLineFromPath = match[1];
        currentFilePatch.diffLineToPath = match[2];
      } else if (line.startsWith(`old mode `)) {
        currentFilePatch.oldMode = line.slice(`old mode `.length).trim();
      } else if (line.startsWith(`new mode `)) {
        currentFilePatch.newMode = line.slice(`new mode `.length).trim();
      } else if (line.startsWith(`deleted file mode `)) {
        currentFilePatch.deletedFileMode = line.slice(`deleted file mode `.length).trim();
      } else if (line.startsWith(`new file mode `)) {
        currentFilePatch.newFileMode = line.slice("new file mode ".length).trim();
      } else if (line.startsWith(`rename from `)) {
        currentFilePatch.renameFrom = line.slice(`rename from `.length).trim();
      } else if (line.startsWith(`rename to `)) {
        currentFilePatch.renameTo = line.slice(`rename to `.length).trim();
      } else if (line.startsWith(`index `)) {
        const match = line.match(/(\w+)\.\.(\w+)/);
        if (!match)
          continue;

        currentFilePatch.beforeHash = match[1];
        currentFilePatch.afterHash = match[2];
      } else if (line.startsWith(`semver exclusivity `)) {
        currentFilePatch.semverExclusivity = line.slice(`semver exclusivity `.length).trim();
      } else if (line.startsWith(`--- `)) {
        currentFilePatch.fromPath = line.slice(`--- a/`.length).trim();
      } else if (line.startsWith(`+++ `)) {
        currentFilePatch.toPath = line.slice(`+++ b/`.length).trim();
      }
    } else {
      if (supportLegacyDiffs && line.startsWith(`--- a/`)) {
        state = `parsing header`;
        commitFilePatch();
        i -= 1;
        continue;
      }

      // parsing hunks
      const lineType = hunkLinetypes[line[0]] || null;
      switch (lineType) {
        case `header`: {
          commitHunk();
          currentHunk = emptyHunk(line);
        } break;

        case null: {
          // unrecognized, bail out
          state = `parsing header`;
          commitFilePatch();
          i -= 1;
        } break;

        case `pragma`: {
          if (!line.startsWith("\\ No newline at end of file"))
            throw new Error(`Unrecognized pragma in patch file: ${line}`);

          if (!currentHunkMutationPart)
            throw new Error(`Bad parser state: No newline at EOF pragma encountered without context`);

          currentHunkMutationPart.noNewlineAtEndOfFile = true;
        } break;

        case `insertion`:
        case `deletion`:
        case `context`: {
          if (!currentHunk)
            throw new Error(`Bad parser state: Hunk lines encountered before hunk header`);

          if (currentHunkMutationPart && currentHunkMutationPart.type !== lineType) {
            currentHunk.parts.push(currentHunkMutationPart);
            currentHunkMutationPart = null;
          }

          if (!currentHunkMutationPart) {
            currentHunkMutationPart = {
              type: lineType,
              lines: [],
              noNewlineAtEndOfFile: false,
            };
          }

          currentHunkMutationPart.lines.push(line.slice(1));
        } break;

        default: {
          miscUtils.assertNever(lineType);
        } break;
      }
    }
  }

  commitFilePatch();

  for (const {hunks} of result)
    if (hunks)
      for (const hunk of hunks)
        verifyHunkIntegrity(hunk);

  return result;
}

export function interpretParsedPatchFile(files: Array<FileDeets>): ParsedPatchFile {
  const result: ParsedPatchFile = [];

  for (const file of files) {
    const {
      semverExclusivity,
      diffLineFromPath,
      diffLineToPath,
      oldMode,
      newMode,
      deletedFileMode,
      newFileMode,
      renameFrom,
      renameTo,
      beforeHash,
      afterHash,
      fromPath,
      toPath,
      hunks,
    } = file;

    const type: PatchFilePart["type"] = renameFrom
      ? `rename`
      : deletedFileMode
        ? `file deletion`
        : newFileMode
          ? `file creation`
          : hunks && hunks.length > 0
            ? `patch`
            : `mode change`;

    let destinationFilePath: string | null = null;
    switch (type) {
      case `rename`: {
        if (!renameFrom || !renameTo)
          throw new Error(`Bad parser state: rename from & to not given`);

        result.push({
          type: `rename`,
          semverExclusivity,
          fromPath: npath.toPortablePath(renameFrom),
          toPath: npath.toPortablePath(renameTo),
        });

        destinationFilePath = renameTo;
      } break;

      case `file deletion`: {
        const path = diffLineFromPath || fromPath;
        if (!path)
          throw new Error("Bad parse state: no path given for file deletion");

        result.push({
          type: `file deletion`,
          semverExclusivity,
          hunk: (hunks && hunks[0]) || null,
          path: npath.toPortablePath(path),
          mode: parseFileMode(deletedFileMode!),
          hash: beforeHash,
        });
      } break;

      case `file creation`: {
        const path = diffLineToPath || toPath;
        if (!path)
          throw new Error(`Bad parse state: no path given for file creation`);

        result.push({
          type: `file creation`,
          semverExclusivity,
          hunk: (hunks && hunks[0]) || null,
          path: npath.toPortablePath(path),
          mode: parseFileMode(newFileMode!),
          hash: afterHash,
        });
      } break;

      case `patch`:
      case `mode change`: {
        destinationFilePath = toPath || diffLineToPath;
      } break;

      default: {
        miscUtils.assertNever(type);
      } break;
    }

    if (destinationFilePath && oldMode && newMode && oldMode !== newMode) {
      result.push({
        type: `mode change`,
        semverExclusivity,
        path: npath.toPortablePath(destinationFilePath),
        oldMode: parseFileMode(oldMode),
        newMode: parseFileMode(newMode),
      });
    }

    if (destinationFilePath && hunks && hunks.length) {
      result.push({
        type: `patch`,
        semverExclusivity,
        path: npath.toPortablePath(destinationFilePath),
        hunks,
        beforeHash,
        afterHash,
      });
    }
  }

  return result;
}

function parseFileMode(mode: string): FileMode {
  const parsedMode = parseInt(mode, 8) & 0o777;

  if (parsedMode !== NON_EXECUTABLE_FILE_MODE && parsedMode !== EXECUTABLE_FILE_MODE)
    throw new Error(`Unexpected file mode string: ${mode}`);

  return parsedMode;
}

export function parsePatchFile(file: string): ParsedPatchFile {
  const lines = file.split(/\n/g);
  if (lines[lines.length - 1] === ``)
    lines.pop();

  try {
    return interpretParsedPatchFile(parsePatchLines(lines, {supportLegacyDiffs: false}));
  } catch (e) {
    if (e instanceof Error && e.message === `hunk header integrity check failed`) {
      return interpretParsedPatchFile(parsePatchLines(lines, {supportLegacyDiffs: true}));
    } else {
      throw e;
    }
  }
}

export function verifyHunkIntegrity(hunk: Hunk) {
  // verify hunk integrity
  let originalLength = 0;
  let patchedLength = 0;

  for (const {type, lines} of hunk.parts) {
    switch (type) {
      case `context`: {
        patchedLength += lines.length;
        originalLength += lines.length;
      } break;

      case "deletion": {
        originalLength += lines.length;
      } break;

      case "insertion": {
        patchedLength += lines.length;
      } break;

      default: {
        miscUtils.assertNever(type);
      } break;
    }
  }

  if (originalLength !== hunk.header.original.length || patchedLength !== hunk.header.patched.length) {
    throw new Error(`hunk header integrity check failed`);
  }
}
