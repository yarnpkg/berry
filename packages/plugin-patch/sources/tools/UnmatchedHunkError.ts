import {Hunk} from './parse';

export class UnmatchedHunkError extends Error {
  constructor(index: number, public hunk: Hunk) {
    super(`Cannot apply hunk #${index + 1}`);
  }
}
