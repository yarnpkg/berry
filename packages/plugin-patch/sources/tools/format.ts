import {miscUtils, Report, formatUtils, MessageName, Configuration} from '@yarnpkg/core';

import {Hunk, PatchMutationType}                                    from './parse';

export function reportHunk(hunk: Hunk, {configuration, report}: {configuration: Configuration, report: Report}) {
  for (const part of hunk.parts) {
    for (const line of part.lines) {
      switch (part.type) {
        case PatchMutationType.Context: {
          report.reportInfo(null, `  ${formatUtils.pretty(configuration, line, `grey`)}`);
        } break;

        case PatchMutationType.Deletion: {
          report.reportError(MessageName.FROZEN_LOCKFILE_EXCEPTION, `- ${formatUtils.pretty(configuration, line, formatUtils.Type.REMOVED)}`);
        } break;

        case PatchMutationType.Insertion: {
          report.reportError(MessageName.FROZEN_LOCKFILE_EXCEPTION, `+ ${formatUtils.pretty(configuration, line, formatUtils.Type.ADDED)}`);
        } break;

        default: {
          miscUtils.assertNever(part.type);
        }
      }
    }
  }
}
