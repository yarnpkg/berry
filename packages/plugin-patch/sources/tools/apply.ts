import {miscUtils, semverUtils}                              from '@yarnpkg/core';
import {FakeFS, ppath, NodeFS, PortablePath}                 from '@yarnpkg/fslib';

import {UnmatchedHunkError}                                  from './UnmatchedHunkError';
import {ParsedPatchFile, FilePatch, Hunk, PatchMutationType} from './parse';

const DEFAULT_TIME = 315532800;

async function preserveTime(baseFs: FakeFS<PortablePath>, p: PortablePath, cb: () => Promise<PortablePath | void>) {
  const stat = await baseFs.lstatPromise(p);

  const result = await cb();
  if (typeof result !== `undefined`)
    p = result;

  if (baseFs.lutimesPromise) {
    await baseFs.lutimesPromise(p, stat.atime, stat.mtime);
  } else if (!stat.isSymbolicLink()) {
    await baseFs.utimesPromise(p, stat.atime, stat.mtime);
  } else {
    throw new Error(`Cannot preserve the time values of a symlink`);
  }
}

export async function applyPatchFile(effects: ParsedPatchFile, {baseFs = new NodeFS(), dryRun = false, version = null}: {baseFs?: FakeFS<PortablePath>, dryRun?: boolean, version?: string | null} = {}) {
  for (const eff of effects) {
    if (eff.semverExclusivity !== null && version !== null)
      if (!semverUtils.satisfiesWithPrereleases(version, eff.semverExclusivity))
        continue;

    switch (eff.type) {
      case `file deletion`: {
        if (dryRun) {
          if (!baseFs.existsSync(eff.path)) {
            throw new Error(`Trying to delete a file that doesn't exist: ${eff.path}`);
          }
        } else {
          await preserveTime(baseFs, ppath.dirname(eff.path), async () => {
            await baseFs.unlinkPromise(eff.path);
          });
        }
      } break;

      case `rename`: {
        if (dryRun) {
          if (!baseFs.existsSync(eff.fromPath)) {
            throw new Error(`Trying to move a file that doesn't exist: ${eff.fromPath}`);
          }
        } else {
          await preserveTime(baseFs, ppath.dirname(eff.fromPath), async () => {
            await preserveTime(baseFs, ppath.dirname(eff.toPath), async () => {
              await preserveTime(baseFs, eff.fromPath, async () => {
                await baseFs.movePromise(eff.fromPath, eff.toPath);
                return eff.toPath;
              });
            });
          });
        }
      } break;

      case `file creation`: {
        if (dryRun) {
          if (baseFs.existsSync(eff.path)) {
            throw new Error(`Trying to create a file that already exists: ${eff.path}`);
          }
        } else {
          const fileContents = eff.hunk
            ? eff.hunk.parts[0].lines.join(`\n`) + (eff.hunk.parts[0].noNewlineAtEndOfFile ? `` : `\n`)
            : ``;

          // Todo: the parent of the first directory thus created will still see its mtime changed
          await baseFs.mkdirpPromise(ppath.dirname(eff.path), {chmod: 0o755, utimes: [DEFAULT_TIME, DEFAULT_TIME]});

          await baseFs.writeFilePromise(eff.path, fileContents, {mode: eff.mode});
          await baseFs.utimesPromise(eff.path, DEFAULT_TIME, DEFAULT_TIME);
        }
      } break;

      case `patch`: {
        await preserveTime(baseFs, eff.path, async () => {
          await applyPatch(eff, {baseFs, dryRun});
        });
      } break;

      case `mode change`: {
        const currentStat = await baseFs.statPromise(eff.path);
        const currentMode = currentStat.mode;

        if (isExecutable(eff.newMode) !== isExecutable(currentMode))
          continue;

        await preserveTime(baseFs, eff.path, async () => {
          await baseFs.chmodPromise(eff.path, eff.newMode);
        });
      } break;

      default: {
        miscUtils.assertNever(eff);
      } break;
    }
  }
}

function isExecutable(fileMode: number) {
  return (fileMode & 0o100) > 0;
}

function trimRight(s: string) {
  return s.replace(/\s+$/, ``);
}

function linesAreEqual(a: string, b: string) {
  return trimRight(a) === trimRight(b);
}

/**
 * How does noNewLineAtEndOfFile work?
 *
 * if you remove the newline from a file that had one without editing other bits:
 *
 *    it creates an insertion/removal pair where the insertion has \ No new line at end of file
 *
 * if you edit a file that didn't have a new line and don't add one:
 *
 *    both insertion and deletion have \ No new line at end of file
 *
 * if you edit a file that didn't have a new line and add one:
 *
 *    deletion has \ No new line at end of file
 *    but not insertion
 *
 * if you edit a file that had a new line and leave it in:
 *
 *    neither insetion nor deletion have the annoation
 *
 */

export async function applyPatch({hunks, path}: FilePatch, {baseFs, dryRun = false}: {baseFs: FakeFS<PortablePath>, dryRun?: boolean}) {
  const mode = await baseFs.statSync(path).mode;

  const fileContents = await baseFs.readFileSync(path, `utf8`);
  const fileLines = fileContents.split(/\n/);

  const result: Array<Array<Modification>> = [];

  let fixupOffset = 0;
  let maxFrozenLine = 0;

  for (const hunk of hunks) {
    const firstGuess = Math.max(maxFrozenLine, hunk.header.patched.start + fixupOffset);

    const maxPrefixFuzz = Math.max(0, firstGuess - maxFrozenLine);
    const maxSuffixFuzz = Math.max(0, fileLines.length - firstGuess - hunk.header.original.length);

    const maxFuzz = Math.max(maxPrefixFuzz, maxSuffixFuzz);

    let offset = 0;
    let location = 0;

    let modifications: ReturnType<typeof evaluateHunk> = null;

    while (offset <= maxFuzz) {
      if (offset <= maxPrefixFuzz) {
        location = firstGuess - offset;
        modifications = evaluateHunk(hunk, fileLines, location);
        if (modifications !== null) {
          offset = -offset;
          break;
        }
      }

      if (offset <= maxSuffixFuzz) {
        location = firstGuess + offset;
        modifications = evaluateHunk(hunk, fileLines, location);
        if (modifications !== null) {
          break;
        }
      }

      offset += 1;
    }

    if (modifications === null)
      throw new UnmatchedHunkError(hunks.indexOf(hunk), hunk);

    result.push(modifications);

    fixupOffset += offset;
    maxFrozenLine = location + hunk.header.original.length;
  }

  if (dryRun)
    return;

  let diffOffset = 0;

  for (const modifications of result) {
    for (const modification of modifications) {
      switch (modification.type) {
        case `splice`: {
          const firstLine = modification.index + diffOffset;
          fileLines.splice(firstLine, modification.numToDelete, ...modification.linesToInsert);
          diffOffset += modification.linesToInsert.length - modification.numToDelete;
        } break;

        case `pop`: {
          fileLines.pop();
        } break;

        case `push`: {
          fileLines.push(modification.line);
        } break;

        default: {
          miscUtils.assertNever(modification);
        } break;
      }
    }
  }

  await baseFs.writeFilePromise(path, fileLines.join(`\n`), {mode});
}

type Push = {
  type: `push`,
  line: string,
};

type Pop = {
  type: `pop`,
};

type Splice = {
  type: `splice`,
  index: number,
  numToDelete: number,
  linesToInsert: Array<string>,
};

type Modification =
  | Push
  | Pop
  | Splice;

function evaluateHunk(hunk: Hunk, fileLines: Array<string>, offset: number): Array<Modification> | null {
  const result: Array<Modification> = [];

  for (const part of hunk.parts) {
    switch (part.type) {
      case PatchMutationType.Context:
      case PatchMutationType.Deletion: {
        for (const line of part.lines) {
          const originalLine = fileLines[offset];

          if (originalLine == null || !linesAreEqual(originalLine, line))
            return null;

          offset += 1;
        }

        if (part.type === PatchMutationType.Deletion) {
          result.push({
            type: `splice`,
            index: offset - part.lines.length,
            numToDelete: part.lines.length,
            linesToInsert: [],
          });

          if (part.noNewlineAtEndOfFile) {
            result.push({
              type: `push`,
              line: ``,
            });
          }
        }
      } break;

      case PatchMutationType.Insertion: {
        result.push({
          type: `splice`,
          index: offset,
          numToDelete: 0,
          linesToInsert: part.lines,
        });

        if (part.noNewlineAtEndOfFile) {
          result.push({type: `pop`});
        }
      } break;

      default: {
        miscUtils.assertNever(part.type);
      } break;
    }
  }

  return result;
}
