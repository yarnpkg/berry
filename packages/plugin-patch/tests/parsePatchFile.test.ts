import {parsePatchFile} from '../sources/tools/parse';

const patch = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,5 +1,5 @@
 this
 is

-a
+
 file
`;
const invalidHeaders1 = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,5 +1,4 @@
 this
 is

-a
+
 file
`;

const invalidHeaders2 = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,4 +1,5 @@
 this
 is

-a
+
 file
`;

const invalidHeaders3 = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,0 +1,5 @@
 this
 is

-a
+
 file
`;
const invalidHeaders4 = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,5 +1,0 @@
 this
 is

-a
+
 file
`;

const invalidHeaders5 = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,5 +1,5@@
 this
 is

-a
+
 file
`;

const accidentalBlankLine = `diff --git a/banana.ts b/banana.ts
index 2de83dd..842652c 100644
--- a/banana.ts
+++ b/banana.ts
@@ -1,5 +1,5 @@
 this
 is

-a
+
 file
`;

const crlfLineBreaks = `diff --git a/banana.ts b/banana.ts
new file mode 100644
index 0000000..3e1267f
--- /dev/null
+++ b/banana.ts
@@ -0,0 +1 @@
+this is a new file
`.replace(/\n/g, `\r\n`);

const modeChangeAndModifyAndRename = `diff --git a/numbers.txt b/banana.txt
old mode 100644
new mode 100755
similarity index 96%
rename from numbers.txt
rename to banana.txt
index fbf1785..92d2c5f
--- a/numbers.txt
+++ b/banana.txt
@@ -1,4 +1,4 @@
-one
+ne

 two

`;

describe(`parsepatchFile`, () => {
  it(`works for a simple case`, () => {
    expect(parsePatchFile(patch)).toMatchSnapshot();
  });

  it(`fails when the patch file has invalid headers`, () => {
    expect(() => parsePatchFile(invalidHeaders1)).toThrow();
    expect(() => parsePatchFile(invalidHeaders2)).toThrow();
    expect(() => parsePatchFile(invalidHeaders3)).toThrow();
    expect(() => parsePatchFile(invalidHeaders4)).toThrow();
    expect(() => parsePatchFile(invalidHeaders5)).toThrow();
  });

  it(`is OK when blank lines are accidentally created`, () => {
    expect(parsePatchFile(accidentalBlankLine)).toEqual(parsePatchFile(patch));
  });

  it(`can handle files with CRLF line breaks`, () => {
    expect(parsePatchFile(crlfLineBreaks)).toMatchSnapshot();
  });

  it(`works`, () => {
    expect(parsePatchFile(modeChangeAndModifyAndRename)).toMatchSnapshot();

    expect(parsePatchFile(accidentalBlankLine)).toMatchSnapshot();
    expect(parsePatchFile(modeChangeAndModifyAndRename)).toMatchSnapshot();
  });
});
