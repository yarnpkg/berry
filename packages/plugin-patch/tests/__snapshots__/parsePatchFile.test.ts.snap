// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsepatchFile can handle files with CRLF line breaks 1`] = `
Array [
  Object {
    "hash": "3e1267f",
    "hunk": Object {
      "header": Object {
        "original": Object {
          "length": 0,
          "start": 1,
        },
        "patched": Object {
          "length": 1,
          "start": 1,
        },
      },
      "parts": Array [
        Object {
          "lines": Array [
            "this is a new file
",
          ],
          "noNewlineAtEndOfFile": false,
          "type": "insertion",
        },
      ],
    },
    "mode": 420,
    "path": "banana.ts",
    "semverExclusivity": null,
    "type": "file creation",
  },
]
`;

exports[`parsepatchFile works 1`] = `
Array [
  Object {
    "fromPath": "numbers.txt",
    "semverExclusivity": null,
    "toPath": "banana.txt",
    "type": "rename",
  },
  Object {
    "newMode": 493,
    "oldMode": 420,
    "path": "banana.txt",
    "semverExclusivity": null,
    "type": "mode change",
  },
  Object {
    "afterHash": "92d2c5f",
    "beforeHash": "fbf1785",
    "hunks": Array [
      Object {
        "header": Object {
          "original": Object {
            "length": 4,
            "start": 1,
          },
          "patched": Object {
            "length": 4,
            "start": 1,
          },
        },
        "parts": Array [
          Object {
            "lines": Array [
              "one",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "deletion",
          },
          Object {
            "lines": Array [
              "ne",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "insertion",
          },
          Object {
            "lines": Array [
              "",
              "two",
              "",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "context",
          },
        ],
      },
    ],
    "path": "banana.txt",
    "semverExclusivity": null,
    "type": "patch",
  },
]
`;

exports[`parsepatchFile works 2`] = `
Array [
  Object {
    "afterHash": "842652c",
    "beforeHash": "2de83dd",
    "hunks": Array [
      Object {
        "header": Object {
          "original": Object {
            "length": 5,
            "start": 1,
          },
          "patched": Object {
            "length": 5,
            "start": 1,
          },
        },
        "parts": Array [
          Object {
            "lines": Array [
              "this",
              "is",
              "",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "context",
          },
          Object {
            "lines": Array [
              "a",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "deletion",
          },
          Object {
            "lines": Array [
              "",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "insertion",
          },
          Object {
            "lines": Array [
              "file",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "context",
          },
        ],
      },
    ],
    "path": "banana.ts",
    "semverExclusivity": null,
    "type": "patch",
  },
]
`;

exports[`parsepatchFile works 3`] = `
Array [
  Object {
    "fromPath": "numbers.txt",
    "semverExclusivity": null,
    "toPath": "banana.txt",
    "type": "rename",
  },
  Object {
    "newMode": 493,
    "oldMode": 420,
    "path": "banana.txt",
    "semverExclusivity": null,
    "type": "mode change",
  },
  Object {
    "afterHash": "92d2c5f",
    "beforeHash": "fbf1785",
    "hunks": Array [
      Object {
        "header": Object {
          "original": Object {
            "length": 4,
            "start": 1,
          },
          "patched": Object {
            "length": 4,
            "start": 1,
          },
        },
        "parts": Array [
          Object {
            "lines": Array [
              "one",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "deletion",
          },
          Object {
            "lines": Array [
              "ne",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "insertion",
          },
          Object {
            "lines": Array [
              "",
              "two",
              "",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "context",
          },
        ],
      },
    ],
    "path": "banana.txt",
    "semverExclusivity": null,
    "type": "patch",
  },
]
`;

exports[`parsepatchFile works for a simple case 1`] = `
Array [
  Object {
    "afterHash": "842652c",
    "beforeHash": "2de83dd",
    "hunks": Array [
      Object {
        "header": Object {
          "original": Object {
            "length": 5,
            "start": 1,
          },
          "patched": Object {
            "length": 5,
            "start": 1,
          },
        },
        "parts": Array [
          Object {
            "lines": Array [
              "this",
              "is",
              "",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "context",
          },
          Object {
            "lines": Array [
              "a",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "deletion",
          },
          Object {
            "lines": Array [
              "",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "insertion",
          },
          Object {
            "lines": Array [
              "file",
            ],
            "noNewlineAtEndOfFile": false,
            "type": "context",
          },
        ],
      },
    ],
    "path": "banana.ts",
    "semverExclusivity": null,
    "type": "patch",
  },
]
`;
